pkg/services/ngalert/notifier/autogen_alertmanager.go

```diff
@@ -0,0 +1,206 @@
+	for receiver := range cfg.ReceiverNames() {
+		setting := models.NewDefaultNotificationSettings(receiver)
+		fp := setting.Fingerprint()
+		notificationSettings[fp] = setting
+	}
+
+	validator := NewNotificationSettingsValidator(cfg)
+	for ruleKey, ruleSettings := range settings {
+		for _, setting := range ruleSettings {
+			// TODO we should register this errors and somehow present to the users or make sure the config is always valid.
+			if err = validator.Validate(setting); err != nil {
+				if skipInvalid {
+					logger.Error("Rule notification settings are invalid. Skipping", append(ruleKey.LogContext(), "error", err)...)
+					continue
+				}
+				return autogeneratedRoute{}, fmt.Errorf("invalid notification settings for rule %s: %w", ruleKey.UID, err)
+			}
+			fp := setting.Fingerprint()
+			// Keep only unique settings.
+			if _, ok := notificationSettings[fp]; ok {
+				continue
+			}
+			notificationSettings[fp] = setting
+		}
+	}
+	if len(notificationSettings) == 0 {
+		return autogeneratedRoute{}, nil
+	}
+	newAutogenRoute, err := generateRouteFromSettings(cfg.GetRoute().Receiver, notificationSettings)
+	if err != nil {
+		return autogeneratedRoute{}, fmt.Errorf("failed to create autogenerated route: %w", err)
+	}
+	return newAutogenRoute, nil
+}
+
+type autogeneratedRoute struct {
+	Route       *definitions.Route
+	Fingerprint data.Fingerprint
+}
+
+// generateRouteFromSettings generates a route and fingerprint for this route. The route is a tree of 3 layers:
+//  1. with matcher by label models.AutogeneratedRouteLabel equals 'true'.
+//  2. with matcher by receiver name.
+//  3. with matcher by unique combination of optional settings. It is created only if there are optional settings.
+func generateRouteFromSettings(defaultReceiver string, settings map[data.Fingerprint]models.NotificationSettings) (autogeneratedRoute, error) {
+	keys := maps.Keys(settings)
+	// sort keys to make sure that the hash we calculate using it is stable
+	slices.Sort(keys)
+
+	rootMatcher, err := labels.NewMatcher(labels.MatchEqual, models.AutogeneratedRouteLabel, "true")
+	if err != nil {
+		return autogeneratedRoute{}, err
+	}
+
+	autoGenRoot := &definitions.Route{
+		Receiver:       defaultReceiver,
+		ObjectMatchers: definitions.ObjectMatchers{rootMatcher},
+		Continue:       false, // We explicitly don't continue toward user-created routes if this matches.
+	}
+
+	receiverRoutes := make(map[string]*definitions.Route)
+	for _, fingerprint := range keys {
+		s := settings[fingerprint]
+		receiverRoute, ok := receiverRoutes[s.Receiver]
+		if !ok {
+			contactMatcher, err := labels.NewMatcher(labels.MatchEqual, models.AutogeneratedRouteReceiverNameLabel, s.Receiver)
+			if err != nil {
+				return autogeneratedRoute{}, err
+			}
+			receiverRoute = &definitions.Route{
+				Receiver:       s.Receiver,
+				ObjectMatchers: definitions.ObjectMatchers{contactMatcher},
+				// We continue on to check all other contact routes.
+				Continue: true,
```

Comment:
not sure I understand why? Can you provide an example of when this can happen?

Reply:
Also, this changes semantics in the case when IsAllDefault() = true and when it's false: in the first case route config implies that it should continue checking other receivers (even though it's not possible) whereas in the latter case, it does not do that.
So, I think we should keep semantics consistent regardless of whether override settings are set or not